Hello Byron!

I’m happy to review your project today.
I will mark your mistakes and give you some hints how it is possible to fix them. We are getting ready for real job, where your team leader/senior colleague will do exactly the same. Don't worry and study with pleasure! 

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Text here.
</div>

# Note to the reviewer, I tried to download my previously reviewed project with your comments, but for some reason I accidentally deleted the file.  I sincerely apologize.

<div class="alert alert-block alert-info">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Thank you for this comment:) I was really confused why your have the second iterations but don't have any comments

</div>

# Sprint 7 Project: Introduction to Machine Learning

# Summary:  This project involves developing a model that could help analyze Megaline subscribers' behavior in regards to whether subscribers were more like to choose the Smart or Ultra Plans.  The types of model being evaluated are: Logistic Regression, Decision Tree and Random Forest.  

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

It's not the best idea to use header format for regular text. Headers should be used only for titles and subtitles.

</div>

### Exploratory Data Analysis and Data Checking/ Cleaning

In [217]:
#Import the necessary libraries.
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import accuracy_score, recall_score

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import confusion_matrix, classification_report

from sklearn.model_selection import GridSearchCV

In [218]:
#Create the dataframe.  Use the try and except clause in order to make sure that the dataframe will upload on a local and non local computer.
try:
    df = pd.read_csv('users_behavior.csv')
except:
    df = pd.read_csv('datasets/users_behavior.csv')

df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [219]:
#Check for duplicates.
df.duplicated().sum()

0

Code Conclusion: 
No duplicates noted in the df dataframe.

In [220]:
#Checking the data types.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Code Conclusion:
All the data types in the dataframe appear to have the correct data type.

In [221]:
#Check for missing values.
df.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

Code Conclusion:
There are no missing values noted in the df dataframe.

### Splitting the Data

In [222]:
#Create a two dataframes, one that contains the features and one that contains the target.
features = df.drop(['is_ultra'], axis=1)

target = df['is_ultra']

In [223]:
#Use the train_test_split() method in order to split the data to a 60/20/20 ratio.
features_train, features_valid_test, target_train, target_valid_test = train_test_split(features, target, test_size=0.4, random_state=135)

features_valid, features_test, target_valid, target_test = train_test_split(features_valid_test, target_valid_test, test_size=0.5, random_state=246)

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct

</div>

Observation:
I have observed that the objective of the task is to analyze/predict whether new Megaline users are more likely to choose either the Smart or the Ultra plan as their carrier.  Since this task is dependent only two distinct outcomes, I can conclude that this is a classification task.  Thus, I chose to pick the: Logistic Regression, Decision Tree and Random Forest Models as possible models best suited to predict the outcome.

# Logistic Regression

### Logistic Regression GridSearch

In [224]:
#Create a dictionary of parameters with lists of possible parameters for GridSearch.
lr_param_grid = {'penalty': ['l1', 'l2'],
                           'C': [0.01, 0.1, 1, 10, 100],
                           'solver': ['liblinear', 'saga']
    
}

In [225]:
#Create a Logistic Regression model using the LogisiticRegression() function from the linear_model  module of the sklearn library.
model = LogisticRegression(random_state=54321)

In [226]:
#Use the GridSearchCV() function in order to create a Gridsearch model in order to choose the best set of parameters for the Logistic Regression model.
lr_gridsearch = GridSearchCV(estimator=model, param_grid=lr_param_grid, cv=5, scoring='accuracy')

#Use the fit() method to train the gridsearch model.
lr_gridsearch.fit(features_train, target_train)

#Print out the best parameters and the best score for the gridsearch model.
print("Logistic Regression Best Parameters:", lr_gridsearch.best_params_)
print("Logisitc Regression Best Score:", lr_gridsearch.best_score_)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\HP\anaconda3\Lib\site-packages\sklearn\linear_model\_sag.py:349: ConvergenceWarning: The max_iter was reached which 

Logistic Regression Best Parameters: {'C': 10, 'penalty': 'l1', 'solver': 'liblinear'}
Logisitc Regression Best Score: 0.7458650158132023


In [246]:
#Enter the parameters from GridSearch into the Logistic Regression model.
model = LogisticRegression(C=10, penalty='l1', solver='liblinear')

#Train the Logistic Regression model using the fit() method.
model.fit(features_train, target_train)

LogisticRegression(C=10, penalty='l1', solver='liblinear')

In [247]:
#Use the score() method in order to calculate the model's accuracy score when the model uses the training set as input.
score_train = model.score(features_train, target_train)

#Use the score() method in order to calculate the model's accuracy score when the model uses the validation set as input.
score_valid = model.score(features_valid, target_valid)

### Logistic Regression Training Set Accuracy Score

In [249]:
#Print the accuracy score of the model's performance on the training set.
print("Logistic Regression Training Set Accuracy Score:", score_train)

Logistic Regression Training Set Accuracy Score: 0.7479253112033195


Conclusion:  
The logistic regression model has an accuracy score of: 0.7479253112033195 when performing tasks on the training set.  This accuracy score is less than the threshold of 0.75 accuracy for this project.  This accuracy score is not satisfactory.

### Logistic Regression Validation Set Accuracy Score

In [251]:
#Print the accuracy score of the model's performance on the training set.
print("Logistic Regression Validation Set Accuracy Score:", score_valid)

Logistic Regression Validation Set Accuracy Score: 0.7527216174183515


Conclusion:
The logistic regression model has an accuracy score of: 0.752721617418385 when performing tasks on the training set.  This accuracy score is greater than the threshold of 0.75 accuracy for this project.  This accuracy score is satisfactory.

### Logistic Regression Prediction Validation Model 

In [252]:
#Use the predict method on the model with the features from the validation set in order to have the model predict the target of a new set of data. 
#This prediction will be used for the confusion matrix. 
prediction_1 = model.predict(features_valid)

### Confusion Matrix Logistic Regression

In [253]:
#Use the confusion_matrix() fucntion from the metrics module of the sklearn library in order to error test the model using the confusion matrix algorithm.
conf_matrix_1 = confusion_matrix(target_valid, prediction_1)

#Print the confusion matrix of the Logistic Regression Model.
print(f"Logisitic Regression Confusion Matrix:\n{conf_matrix_1}")

Logisitic Regression Confusion Matrix:
[[436   9]
 [150  48]]


Logistic Regression Confusion Matrix Conclusion:
According to the confusion matrix, the logistic regression model generated a total of 159 errors for this task.

### Classificaion Report Logistic Regression

In [254]:
class_report_1 = classification_report(target_valid, prediction_1)

print(f"Logistic Regression Classification Report:\n{class_report_1}")

Logistic Regression Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.98      0.85       445
           1       0.84      0.24      0.38       198

    accuracy                           0.75       643
   macro avg       0.79      0.61      0.61       643
weighted avg       0.77      0.75      0.70       643



# Decision Tree

### Decision Tree Gridsearch

In [233]:
#Create a dictionary that contains lists of possible parameters for the Decision Tree model.
tree_param_grid ={
    'max_depth' : [2, 4, 6, 8, 10, 12, 15],
    'min_samples_split': [2, 5, 10, 20],
    'max_features': ['auto', 'sqrt', 'log2'],
    'criterion': ['gini', 'entropy']}

In [234]:
#Create a Decision Tree model using the DecisionTreeClassifier() function from the tree module of the sklearn library.
model_2 = DecisionTreeClassifier(random_state=123)

In [235]:
#Use the GridsearchCV() function in order to find the set of best parameters for the Decision Tree model.
tree_grid_search = GridSearchCV(estimator=model_2, param_grid=tree_param_grid, cv=5, scoring='accuracy')

#Use the fit() method in order to train the GridSearch model.
tree_grid_search.fit(features_train, target_train)

#Print the best set of parameteres and the best score for the Decision Tree model.
print("Decision Tree Best Parameters:", tree_grid_search.best_params_)
print("Decison Tree Best Score:", tree_grid_search.best_score_)

Decision Tree Best Parameters: {'criterion': 'gini', 'max_depth': 8, 'max_features': 'sqrt', 'min_samples_split': 5}
Decison Tree Best Score: 0.795649014198237


c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
280 fits failed out of a total of 840.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
280 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\HP\anaconda3\Lib\site-packages\sklearn\base.py", line 1466, in wrapper
    estimator._validate_params()
  File "c:\Users\HP\anaconda3\Lib\site-packages\sklearn\base.py", line 666, in _validate_params
    validate_parameter_constraints(
  File "c:\Users\HP\anaconda3\Lib\site-packages\sklearn\util

### Decision Tree Training Set Accuracy Score

In [257]:
#Enter the parameters from GridSearch into the Decsion Tree model.
model_2 = DecisionTreeClassifier(criterion='gini', max_depth=8, max_features='sqrt', min_samples_split=5, random_state=123)

In [258]:
#Train the Decision Tree model using the fit() method.
model_2.fit(features_train, target_train)

DecisionTreeClassifier(max_depth=8, max_features='sqrt', min_samples_split=5,
                       random_state=123)

In [259]:
#Use the score() method in order to calculate the Decision Tree model's accuracy score when using the training set.
score_2_train = model_2.score(features_train, target_train)

#Print the accuracy score.
print("Decsion Tree Training Set Accuracy Score:", score_2_train)

Decsion Tree Training Set Accuracy Score: 0.841804979253112


Conclusion:
The accuracy score of the Decision Tree model when performing on the training set is 0.841804979253112.  The accuracy score of the Decision Tree is greater than the accuracy score of 0.75 for the threshold.  This accuracy score is satisfactory.

### Decision Tree Validation Set Accuracy Score

In [260]:
#Use the score() method to calculate the Decision Tree model's accuracy score when using the validation set.
score_2 = model_2.score(features_valid, target_valid)

#Print the accuracy score.
print("Decision Tree Validation Set Accuracy Score:", score_2)

Decision Tree Validation Set Accuracy Score: 0.7962674961119751


Conclusion:
The accuracy score of the Decision Tree model when performing on the validation set is 0.796267491119751.  The accuracy score of the Decision Tree is greater than the accuracy score of 0.75 for the threshold.  This accuracy score is satisfactory.

### Decision Tree Model Prediction

In [261]:
#Use the predict method on the model with the features from the validation set in order to have the model predict the target of a new set of data. 
#This prediction will be used for the confusion matrix. 
prediction_2 = model_2.predict(features_valid)

### Decision Tree Confusion Matrix

In [262]:
#Use the confusion_matrix() fucntion from the metrics module of the sklearn library in order to error test the model using the confusion matrix algorithm.
conf_matrix_2 = confusion_matrix(target_valid, prediction_2)

#Print the confusion matrix of the Decision Tree Model.
print(f"Decision Tree Confusion Matrix:\n{conf_matrix_2}")

Decision Tree Confusion Matrix:
[[424  21]
 [110  88]]


Conclusion:
According to the confusion matrix, the Decision Tree model generated a total of 131 errors.

### Decision Tree Classification Report

In [263]:
class_report_2 = classification_report(target_valid, prediction_2)

print(f"Decision Tree Classifcation Report:\n{class_report_2}")

Decision Tree Classifcation Report:
              precision    recall  f1-score   support

           0       0.79      0.95      0.87       445
           1       0.81      0.44      0.57       198

    accuracy                           0.80       643
   macro avg       0.80      0.70      0.72       643
weighted avg       0.80      0.80      0.78       643



# Random Forest

In [242]:
#Create a Random Forest model using the RandomForestClassifier() method from the ensemble module of the skleanlibrary.  GridSearch was not able to
#generate an output for this model, so I decided to not enter any specific parameters.
model_3 = RandomForestClassifier(random_state=818)

In [264]:
#Use the fit() method in order to train the Random Forest model.
model_3.fit(features_train, target_train)

RandomForestClassifier(random_state=818)

### Random Forest Training Accuracy Score

In [265]:
#Use the score method to calculate the accuracy score for the Random Forest model when working with the training set.
score_3_train = model_3.score(features_train, target_train)

#Print the accuracy score.
print("Random Forest Training Set Accuracy Score:", score_3_train)

Random Forest Training Set Accuracy Score: 1.0


Conclusion:
The accuracy score of the Random Forest model after performing with the data from the training set is: 1.0.  The Random Forest model's accuracy score is greater than the threshold accuracy score which is 0.75.  This accuracy score is satisfactory.

### Random Forest Validation Test Accuracy Score

In [266]:
#Calculate the accuracy score of the Random Forest model when working with the validation set.
score_3 = model_3.score(features_valid, target_valid)

#Print the accuracy score.
print("Random Forest Validation Set Accuracy Score:", score_3)

Random Forest Validation Set Accuracy Score: 0.80248833592535


Conclusion:
The accuracy score of the Random Forest model after performing with the data from the validation set is: 0.80248833592535.  The Random Forest model's accuracy score is greater than the threshold accuracy score which is 0.75. This accuracy score is satisfactory.

### Random Forest Model Prediction

In [267]:
#Use the predict method on the model with the features from the validation set in order to have the model predict the target of a new set of data. 
#This prediction will be used for the confusion matrix. 
prediction_3 = model_3.predict(features_valid)

### Random Forest Confusion Matrix

In [268]:
#Use the confusion_matrix() fucntion from the metrics module of the sklearn library in order to error test the model using the confusion matrix algorithm.
conf_matrix_3 = confusion_matrix(target_valid, prediction_3)

#Print the confusion matrix of the random forest model.
print(f"Random Forest Confusion Matrix:\n{conf_matrix_3}")

Random Forest Confusion Matrix:
[[403  42]
 [ 85 113]]


Conclusion:
The Random Forest model generated a total of 97 Errors.

### Random Forest Classification Report

In [269]:
class_report_3 = classification_report(target_valid, prediction_3)

print(f"Random Forest Classification Report:\n{class_report_3}")

Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.91      0.86       445
           1       0.73      0.57      0.64       198

    accuracy                           0.80       643
   macro avg       0.78      0.74      0.75       643
weighted avg       0.80      0.80      0.80       643



# Choosing The Best Model

Decision:
According to the confusion matrices, I have decided to choose the Random Forest model as the best model to perform predictions for this task.
The Random Forest model generated a total of 97 errors, which is less errors than the Logistic Regression model (159 errors) and the Decision Tree model (131 errors). I will now proceed to calculate the accuracy score of the Random Forest model when working with the testing set data.  The Random Forest model also had good values for precision, recall, f1-score, support and accuracy according to the classification reports as well.

### Decision Tree Test

In [270]:
#Using the score() method, calculate the accuracy score of the Decision Tree Model when working with the testing set.
score_test = model_3.score(features_test, target_test)

#Print the accuracy score.
print(f"Random Forest Test Accuracy: {score_test}")

Random Forest Test Accuracy: 0.8055987558320373


Conclusion:
The accuracy score of the Decision Tree Model when working with the data from the test set it 0.8055987558320373.  This accuracy score is greater than accuracy score of the project's threshold, 0.75.  This accuracy score is satisfactory.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

I don't know what was incorrect previous time, but now all is fine:) Well done!

</div>

# General Conclusion:
Overall, the Random Forest model appeared to be the best model to solve the task in order to predict whether Megaline users were more likely to choose the Smart or Ultra plans.  The Random Forest model's accuracy scores when working with the training and validation sets were above the project's 0.75 accuracy score threshold.  The Random Forest model also had good values for precision, recall, f1-score, support and accuracy according to the classification report as well.  The main determining factor as to why I chose the Random Model as the best model for the task is because the confusion matrix determined that the Random Forest model only generated 97 errors when making predictions with the validation set, where as the Logistic Regression model generated 159 errors and the Decision Tree model generated 131 errors.  Finally, the Random Forest model's accuracy score when working with the testing set was greater than the 0.75 accuracy score threshold required for this project, thus making the Random Forest model a good model for predicting whether Megaline users were more likely to choose the Smart or Ultra plans in the future.